### PR TITLE
Final_project: orm_theority 

### PR DESCRIPTION
**Pull request by Katerina-Charakhovich**

### CODE REVIEW

**Comment by ShiniTada** [comment]
add info about cache (popular interview topic)
add info about n+1 problem

File: stage #2.5/module #2 ORM - JPA, Hibernate/#2. Defining JPA Entities/content.md [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+Each entity is associated with some metadata that represents the information of it. Instead of database, this metadata
+is existed either inside or outside the class. This metadata can be in following forms: 
+
+* **Annotation** - In Java, annotations are the form of tags that represents metadata. This metadata persist inside the class.
+* **XML** - In this form, metadata persist outside the class in XML file.
+
+
+#####2.2. The Entity annotations
+All the JPA annotations are defined in the **javax.persistence** package. HHibernate Annotations are based on the JPA 2
```
Comment (ShiniTada) [inline stage #2.5/module #2 ORM - JPA, Hibernate/#2. Defining JPA Entities/content.md:22]: HHibernate

File: stage #2.5/module #2 ORM - JPA, Hibernate/#2. Defining JPA Entities/content.md [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+      private String name;
+        
+         // other fields, getters and setters
+     }
+The name element specifies the name of the column in the table. The length element specifies its length. The 
+nullable element specifies whether the column is nullable or not, and the unique element specifies whether the column
+is unique. If we don't specify this annotation, the name of the field will be considered the name of the column in the table.
+
+#####2.2.5. The Transient Annotation
+Sometimes, we may want to make a field non-persistent. We can use the **@Transient**  annotationto do so. It specifies
```
Comment (ShiniTada) [inline stage #2.5/module #2 ORM - JPA, Hibernate/#2. Defining JPA Entities/content.md:99]: annotationto -> annotation to

File: stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+persistent entity instances, to find entities by their primary key identity, and to query over all entities. This
+interface is similar to the Session in Hibernate. 
+
+**SessionFactory** (org.hibernate.SessionFactory)</br>
+A thread-safe (and immutable) representation of the mapping of the application domain model to a database. Acts as a 
+factory for org.hibernate.Session instances. The EntityManagerFactory is the JPA equivalent of a SessionFactory and 
+basically, those two converge into the same SessionFactory implementation.
+A SessionFactory is very expensive to create, so, for any given database, the application should have only one 
+associated SessionFactory. The SessionFactory maintains services that Hibernate uses across all Session(s) such as s
+econd level caches, connection pools, transaction system integrations, etc.
```
Comment (ShiniTada) [inline stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md:52]: s econd -> second

File: stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+* REFRESH - cascades the entity refresh operation.
+* DETACH - cascades the entity detach operation.
+
+Additionally, the CascadeType.ALL will propagate any Hibernate-specific operation, which is defined by the 
+org.hibernate.annotations.CascadeType enum:
+* SAVE_UPDATE - cascades the entity saveOrUpdate operation.
+* REPLICATE - cascades the entity replicate operation.
+* LOCK - cascades the entity lock operation.
+
+The following examples will explain some aforementioned cascade operations using the following entities:
```
Comment (ShiniTada) [inline stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md:243]: aforementioned -> before mentioned

File: stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    Decimal amount;
+    @ManyToOne (fetch = FetchType.LAZY)
+    private Customer customer;
+    ...
+    }
+
+###3.1.2. Container and Application Managed EntityManager
+Basically, there are two types of EntityManager: Container-Managed and Application-Managed.
+####3.1.2.1. Container-Managed EntityManager
+Here, the container((such as a JEE Container or Spring) ) injects the EntityManager in enterprise components.
```
Comment (ShiniTada) [inline stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md:270]: ((such as a JEE Container or Spring) ) -> (such as a JEE Container or Spring)

File: stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+This also means the container is in charge of beginning the transaction, as well as committing or rolling it back.
+
+####3.1.2.2. Application-Managed EntityManager
+An application-managed entity manager allows you to control the entity manager in application code.
+This entity manager is retrieved through the EntityManagerFactory API. In order to create an EntityManager,
+we must explicitly call createEntityManager() in the EntityManagerFactory:
+
+    protected static EntityManagerFactory entityManagerFactory;
+    static {
+        entityManagerFactory = Persistence.createEntityManagerFactory("com.epam");
```
Comment (ShiniTada) [inline stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md:290]: "com.epam" -> "com.mjc"

File: stage #2.5/module #2 ORM - JPA, Hibernate/#2. Defining JPA Entities/content.md (hunk 1 above)
Comment (Katerina-Charakhovich) [reply stage #2.5/module #2 ORM - JPA, Hibernate/#2. Defining JPA Entities/content.md:22]: done

File: stage #2.5/module #2 ORM - JPA, Hibernate/#2. Defining JPA Entities/content.md (hunk 2 above)
Comment (Katerina-Charakhovich) [reply stage #2.5/module #2 ORM - JPA, Hibernate/#2. Defining JPA Entities/content.md:99]: done

File: stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md (hunk 3 above)
Comment (Katerina-Charakhovich) [reply stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md:52]: done

File: stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md (hunk 4 above)
Comment (Katerina-Charakhovich) [reply stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md:243]: done

File: stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md (hunk 6 above)
Comment (Katerina-Charakhovich) [reply stage #2.5/module #2 ORM - JPA, Hibernate/#3. Persistence Context/content.md:290]: done

**Comment by ShiniTada** [comment]
Please rename root package to 'module #3 ORM - JPA, Hibernate'

**Comment by ShiniTada** [comment]
please also update module_description file with listing chapters. Example:
https://github.com/mjc-school/MJC-School/blob/final_project/stage%20%232.5/module%20%231%20Architecture%20advanced/module_description.md

File: stage #2.5/module #2 ORM - JPA, Hibernate/#6. Hibernate Caching/content.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    <property name="hibernate.cache.use_query_cache" value="true"/>
+
+Once done, we need to explicitly call **_setCacheable(true)_** on Query object before execution:
+
+    Query query = session.createQuery("select c from Customer c where c.name = ?");
+    query.setCacheable(true);
+
+The query cache is similar to the second level cache. But unlike it, the key to the cache data is not the object 
+identifier, but the set of query parameters. And the data itself is the identifiers of the objects that match the query 
+criteria. Thus, it is rational to use this cache with the second-level cache.
```
Comment (ShiniTada) [inline stage #2.5/module #2 ORM - JPA, Hibernate/#6. Hibernate Caching/content.md:108]: Why don't you use picture in media folder?

**Comment by ShiniTada** [comment]
I reviewed new changes. Good job, great simple explanations. I appreciate it!

File: stage #2.5/module #2 ORM - JPA, Hibernate/#6. Hibernate Caching/content.md (hunk 7 above)
Comment (Katerina-Charakhovich) [reply stage #2.5/module #2 ORM - JPA, Hibernate/#6. Hibernate Caching/content.md]: corrected

**Pull request merged**